### PR TITLE
Add support for defmt 0.3

### PR DESCRIPTION
**Pull request by newAM**
Update defmt (and other knurling-rs tools) to 0.3, see <https://ferrous-systems.com/blog/defmt-3/>

### CODE REVIEW

**Review by talhaHavadar — COMMENTED** [review]
Looks ok to me, but there is a failure in build.

**Comment by korken89** [comment]
This is how I'm solving it in `heapless` to support both 0.2 and 0.3: https://github.com/japaric/heapless/pull/251
Might be of interest.

**Comment by newAM** [comment]
> This is how I'm solving it in `heapless` to support both 0.2 and 0.3: [japaric/heapless#251](https://github.com/japaric/heapless/pull/251) Might be of interest.

That's a good idea, I updated it to use a range.

> Looks ok to me, but there is a failure in build.

I cannot reproduce the same failure on Linux, and unfortunately I don't have a Windows install easily accessible right now.

Are you able to help me out with this?

**Comment by jonas-schievink** [comment]
please ping me once CI is fixed

**Comment by newAM** [comment]
Somebody on matrix pointed out the Windows build is not building for an embedded target, which is not supported by defmt: https://github.com/knurling-rs/defmt/issues/463

I fixed the failure by removing that build for Windows :sweat_smile:.

@jonas-schievink CI is fixed.

**Comment by jonas-schievink** [comment]
hmm, why did it work with 0.2?

**Comment by newAM** [comment]
> hmm, why did it work with 0.2?

I did a quick scan through the code.

`defmt` 0.3 introduced some new global symbols for flushing:

https://github.com/knurling-rs/defmt/blob/124a6f6f3abe4e77e885d6fab001872a2da08ee8/defmt/src/lib.rs#L377-L382

These symbols get defined by the global logger:

https://github.com/knurling-rs/defmt/blob/8a6e8eebe40f943d9b0ba8725cd6da033b9c399e/macros/src/lib.rs#L24-L28

https://github.com/knurling-rs/defmt/blob/e240ba99bc0c2258bc8ba2d7e4694e568f19c758/macros/src/attributes/global_logger.rs#L41-L45

I am guessing there is something about Windows that doesn't like GNU linker script or the `export_name` attributes that were introduced for these new symbols.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -35,9 +35,10 @@ jobs:
         toolchain: ${{ matrix.rust }}
         override: true
     - name: Build
-      run: |
-        cargo build --all-targets
-        cargo build --all-targets --all-features
+      run: cargo build --all-targets
+    - name: Build All Features
+      if: ${{ matrix.os != 'windows-latest' }}
```
Comment (jonas-schievink) [inline .github/workflows/ci.yml:40]:
```suggestion
    - name: Build All Features
      # defmt 0.3 doesn't build when targeting windows
      if: ${{ matrix.os != 'windows-latest' }}
```

**Comment by newAM** [comment]
Thanks for the review & merge!

Would it be possible to get a `0.6.1` release with this change?

**Comment by jonas-schievink** [comment]
Published 0.6.1

Regarding the linker error, I don't think that's a problem with defmt 0.3.0 specifically, since https://github.com/stm32-rs/bxcan/pull/49 also fails to link again defmt 0.2.3 (https://github.com/stm32-rs/bxcan/runs/4213689874?check_suite_focus=true)

**Pull request merged**